File: 01- January/10- Same Tree/10- Same Tree (Noura Algohary).py
# Author: Noura Algohary
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # Both nodes are NULL
        if(not p and not q):
            return True
        
        # One of the nodes is NULL
        if (not p or not q):
            return False
        
        # Both nodes exist but different numbers
        if (p.val != q.val):
            return False

        # Apply the same function on all nodes from left to right
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

File: 01- January/10- Same Tree/10- Same Tree (Noura Algohary).cpp
// Author: Noura Algohary
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // Both nodes are NULL
        if(!p && !q) return true;

        // One of the nodes is NULL
        if (!p || !q) return false;
        
        // Both nodes exist but different numbers
        if(p->val != q->val) return false;
        
        // Apply the same function on all nodes from left to right
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};